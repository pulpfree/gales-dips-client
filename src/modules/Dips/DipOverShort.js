import React, { Component } from 'react'
import PropTypes from 'prop-types'

import classNames from 'classnames'

import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'

import { fmtNumber } from '../../utils/utils'

// fixme: I think this should be a pure component
class DipOverShort extends Component {

  render() {

    const { classes, data } = this.props

    let rows = []
    let haveData = false
    let fuelPrice
    let displayDate

    if (data && data.loading === false && data.dipOverShortRange && data.dipOverShortRange[1]) {
      haveData = true
      const os = data.dipOverShortRange[1].overShort
      for (const prop in os) {
        rows.push({
          fuelType:   os[prop].fuelType,
          litres:     os[prop].litres,
          overshort:  os[prop].overshort,
          sale:       os[prop].sale,
        })
      }
    }

    if (haveData && data.fuelPrice) {
      fuelPrice = fmtNumber(data.fuelPrice.price, 2)
      displayDate = this.props.dateObj.format('ddd, MMM Do')
    }

    return (
      <div className={classes.container}>
        <Typography
            gutterBottom
            variant="title"
        >Overshort</Typography>
        {haveData &&
          <div className={classes.container}>
            <div className={classes.headerRow}>
              <div className={classes.headerCell}>Fuel Type</div>
              <div className={classNames([classes.headerCell], [classes.alignRight])}>Dip Litres</div>
              <div className={classNames([classes.headerCell], [classes.alignRight])}>Sale Litres</div>
              <div className={classNames([classes.headerCell], [classes.alignRight])}>Over/Short</div>
            </div>

            {rows.map((os, i) => (
              <div
                  className={classes.dataRow}
                  key={i}
              >
                <div className={classes.dataCell}>{os.fuelType}</div>
                <div className={classNames([classes.dataCell], [classes.alignRight])}>{fmtNumber(os.litres, 3)}</div>
                <div className={classNames([classes.dataCell], [classes.alignRight])}>{fmtNumber(os.sale, 3)}</div>
                <div className={classNames([classes.dataCell], [classes.alignRight], {[classes.negative]: os.overshort < 0})}>{fmtNumber(os.overshort, 3)}</div>
              </div>
            ))}
          </div>
        }
        {fuelPrice &&
        <div className={classes.fuelPrice}>
        <Typography
            gutterBottom
            variant="subheading"
        >Fuel Price: {fuelPrice} - ({displayDate})</Typography>
        </div>
        }
      </div>
    )
  }
}

DipOverShort.propTypes = {
  classes:  PropTypes.object.isRequired,
  data:     PropTypes.object,
  dateObj:  PropTypes.object,
}

const styles =  theme => ({
  alignRight: {
    textAlign: 'right',
  },
  alignCenter: {
    textAlign: 'center',
  },
  container: {
    display:        'flex',
    flexDirection:  'column',
    fontFamily:     theme.typography.fontFamily,
    marginLeft:     theme.spacing.unit * 6,
    width:          theme.spacing.unit * 55,
  },
  dataCell: {
    alignSelf:      'flex-end',
    flex:           1,
    padding:        theme.spacing.unit,
  },
  dataRow: {
    borderBottomColor:  '#efefef',
    borderBottomStyle:  'solid',
    borderBottomWidth:  1,
    display:            'inline-flex',
    flexDirection:      'row',
    '&:hover': {
      backgroundColor: theme.palette.grey['50'],
    },
  },
  fuelPrice: {
    marginTop: theme.spacing.unit * 2,
  },
  headerRow: {
    borderBottomColor:  '#efefef',
    borderBottomStyle:  'solid',
    borderBottomWidth:  1,
    display:            'inline-flex',
    flexDirection:      'row',
    paddingBottom:      4,
  },
  headerCell: {
    color:      theme.palette.secondary.main,
    flex:       1,
    fontWeight: '500',
    padding:    theme.spacing.unit,
  },
  negative: {
    color: 'red',
  },
})

export default withStyles(styles)(DipOverShort)
