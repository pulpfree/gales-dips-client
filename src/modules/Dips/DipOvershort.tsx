import React from 'react'

import { Paper, Table, TableBody, TableHead, TableRow, TableCell, TableContainer, Typography } from '@material-ui/core'
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles'
import classNames from 'classnames'
import { format, parse } from 'date-fns'

import { FuelTypes, OvershortData } from './types'
import { fmtNumber } from '../../utils'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    container: {
      margin: theme.spacing(1),
      marginTop: theme.spacing(2),
    },
    fuelPriceRow: {
      margin: theme.spacing(1),
      marginTop: theme.spacing(2),
    },
    negative: {
      color: theme.palette.primary.main,
    },
  }),
)

interface OSProps {
  data: OvershortData
}

export const DipOvershort = ({ data }: OSProps): JSX.Element => {
  const classes = useStyles()

  let fuelPrice
  let displayDate

  const fts = Object.values(FuelTypes)
  const oss = data.overshort.overShort
  const rows = []
  if (oss) {
    for (let i = 0; i < fts.length; i++) {
      if (oss[fts[i]]) {
        rows.push(oss[fts[i]])
      }
    }
  }
  if (data.fuelPrice) {
    fuelPrice = fmtNumber(data.fuelPrice.price, 2)
    displayDate = format(parse(data.fuelPrice.date.toString(), 'yyyyMMdd', new Date()), 'E, MMM Do yyyy')
  }
  return (
    <>
      <Typography variant='h6'>Overshort</Typography>
      <TableContainer component={Paper}>
        <Table aria-label='overshort table' size='small'>
          <TableHead>
            <TableRow>
              <TableCell>Fuel Type</TableCell>
              <TableCell align='right'>Dip Litres</TableCell>
              <TableCell align='right'>Litres Sold</TableCell>
              <TableCell align='right'>Overshort</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {rows.map((os) => (
              <TableRow key={os.fuelType}>
                <TableCell>{os.fuelType}</TableCell>
                <TableCell align='right'>{fmtNumber(os.tankLitres, 3, true)}</TableCell>
                <TableCell align='right'>{fmtNumber(os.litresSold, 3, true)}</TableCell>
                <TableCell align='right' className={classNames({ [classes.negative]: os.overShort < 0 })}>
                  {fmtNumber(os.overShort, 3, true)}
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
      <div className={classes.fuelPriceRow}>
        Fuel Price: {fuelPrice} ({displayDate})
      </div>
    </>
  )
}
