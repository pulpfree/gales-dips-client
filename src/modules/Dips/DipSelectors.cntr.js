import * as errorActions from '../Error/errorActions'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { compose } from 'react-apollo'
import { withRouter } from 'react-router'

import DipSelectors from './DipSelectors'

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...errorActions,
    }, dispatch),
  }
}

export default compose(
  withRouter,
  connect(
    null,
    mapDispatchToProps
  ),
)(DipSelectors)
