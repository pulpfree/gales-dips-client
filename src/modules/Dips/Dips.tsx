import React from 'react'

import { Grid, Icon, IconButton, FormControl } from '@material-ui/core'
import { KeyboardDatePicker } from '@material-ui/pickers'
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles'
import { addDays, subDays, format } from 'date-fns'
import { useHistory, useParams } from 'react-router-dom'
import { utcToZonedTime } from 'date-fns-tz'

import { ContentContainer } from '../Base/ContentContainer'
import { HUMAN_DATE_FORMAT, STD_DATE_FORMAT } from '../../config'
import { ImportDataLink } from './ImportDataLink'
import { Loader } from '../Base/Loader'
import { PageTitle } from '../Base/PageTitle'
import { StationSelector } from '../Base/StationSelector'
import { useAlertDispatch } from '../Base/Alert'
import { useDips } from './useDips'
import { DipsForm } from './DipsForm'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    navButton: {
      // backgroundColor: theme.palette.secondary.light,
      // maxWidth: 60,
      // marginLeft: theme.spacing(1),
      // marginRight: theme.spacing(1),
    },
    navIcon: {
      // margin: -4,
      // paddingLeft: 4,
      // paddingRight: -60,
    },
    navControl: {
      flexDirection: 'row',
      marginTop: theme.spacing(1),
    },
    formControl: {
      margin: theme.spacing(1),
      minWidth: 120,
    },
    textField: {
      marginLeft: theme.spacing(1),
      marginRight: theme.spacing(1),
      width: 200,
    },
  }),
)

export function Dips(): JSX.Element {
  const classes = useStyles()
  const history = useHistory()
  const alertDispatch = useAlertDispatch()
  const [nextDisabled, setNextDisabled] = React.useState<boolean>(true)
  const { date: pDate, stationID: pStationID } = useParams<{ date: string; stationID: string }>()
  const {
    date,
    dipTankData,
    error,
    haveCurrentDips,
    loading,
    fuelSaleDate,
    stationID,
    setDate,
    setStationID,
  } = useDips()

  const handleNextPrevDate = (val: string) => {
    const dte = date as Date
    if (val === 'p') {
      setDate(subDays(dte, 1))
    } else if (val === 'n') {
      setDate(addDays(dte, 1))
    }
  }

  // Check for url parameters and set
  React.useEffect(() => {
    if (pDate && pStationID) {
      setDate(utcToZonedTime(new Date(pDate), 'Europe/London'))
      setStationID(pStationID)
    }
  }, [pDate, pStationID, setDate, setStationID])

  const dte = date as Date
  React.useEffect(() => {
    const handleSetNextDisable = () => {
      const selectDteFmt = Number(format(dte, 'yyyyMMdd'))
      const todayFmt = Number(format(new Date(), 'yyyyMMdd'))
      const isPast = selectDteFmt >= todayFmt
      setNextDisabled(isPast)
    }
    handleSetNextDisable()
    if (stationID) {
      const fmtDte = format(dte, STD_DATE_FORMAT)
      const uri = `/dips/${fmtDte}/${stationID}`
      history.push(uri)
    }
  }, [dte, history, stationID])

  React.useEffect(() => {
    if (error) {
      alertDispatch.send({ message: error, severity: 'error', setTimeout: false, title: 'Data Error' })
    }
  }, [alertDispatch, error])

  return (
    <ContentContainer>
      <PageTitle title='Dip Entries' />

      <Grid container spacing={3} style={{ maxWidth: 550 }}>
        <Grid item xs={4}>
          <FormControl className={classes.formControl}>
            <StationSelector stationHandler={setStationID} />
          </FormControl>
        </Grid>

        <Grid item xs={4}>
          <FormControl className={classes.formControl} style={{ width: 160 }}>
            <KeyboardDatePicker
              autoOk
              disableFuture
              format={HUMAN_DATE_FORMAT}
              label='Date'
              onChange={setDate}
              value={date}
            />
          </FormControl>
        </Grid>

        <Grid item xs={4}>
          <FormControl className={classes.navControl}>
            <IconButton className={classes.navButton} onClick={() => handleNextPrevDate('p')}>
              <Icon className={classes.navIcon} fontSize='large'>
                arrow_back_ios
              </Icon>
            </IconButton>
            <IconButton className={classes.navButton} disabled={nextDisabled} onClick={() => handleNextPrevDate('n')}>
              <Icon className={classes.navIcon} fontSize='large'>
                arrow_forward_ios
              </Icon>
            </IconButton>
          </FormControl>
        </Grid>
      </Grid>
      {loading && <Loader />}
      {!haveCurrentDips && !loading && fuelSaleDate && <ImportDataLink date={fuelSaleDate} />}
      {haveCurrentDips && !loading && dipTankData && <DipsForm data={dipTankData} />}
    </ContentContainer>
  )
}
