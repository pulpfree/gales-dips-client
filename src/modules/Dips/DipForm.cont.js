import gql from 'graphql-tag'
import { compose, graphql, withApollo } from 'react-apollo'
import { connect } from 'react-redux'


import { bindActionCreators } from 'redux'

import DipForm from './DipForm.cont'
import * as errorActions from '../Error/errorActions'


const CREATE_DIPS = gql`
mutation CreateDips($fields: [DipInput]) {
  createDips(input: $fields) {
    ok
    nModified
  }
}
`

const PersistDip = graphql(CREATE_DIPS, {
  props: ({ mutate }) => ({
    UpdateDip: (fields) => mutate({
      variables: {fields},
      errorPolicy: 'all',
    }),
  }),
  options: (props) => ({
    /*onCompleted: () => {
    // onCompleted: (result, errs) => {
      // console.log('result: ', result)
      props.history.push('/admin/tank-admin')
    },*/
    // refetchQueries: [{query: TANKLIST_QUERY}],
  }),
  errorPolicy: 'all',
})

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...errorActions,
    }, dispatch),
  }
}

export default compose(
  withApollo,
  PersistDip,
  connect(
    null,
    mapDispatchToProps
  ),
)(DipForm)