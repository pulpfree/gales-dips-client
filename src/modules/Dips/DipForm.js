import React, { Component } from 'react'
import PropTypes from 'prop-types'

import gql from 'graphql-tag'
import { Mutation } from 'react-apollo'

import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import classNames from 'classnames'
import { debounce } from 'lodash'

import Button from '@material-ui/core/Button'
import Input from '@material-ui/core/Input'
import Save from '@material-ui/icons/Save'
import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'

import Toaster from '../Common/Toaster'
import * as errorActions from '../Error/errorActions'
import Alert from '../Common/Alert'
import { fmtNumber } from '../../utils/utils'
import { DIP_QUERY } from './Dips'
import { FUEL_TYPE_LIST as fuelTypeList } from '../../config/constants'
const R = require('ramda')


const CREATE_DIPS = gql`
mutation CreateDips($fields: [DipInput]) {
  createDips(input: $fields) {
    ok
    nModified
  }
}
`

class DipForm extends Component {

  constructor(props) {
    super(props)
    this.state = {
      tanks: null,
      toasterMsg: '',
    }
    // this.createDipsParams = this.createDipsParams.bind(this)
    // this.handleChange = this.handleChange.bind(this)
    this.handleCalculateLitres = debounce(this.handleCalculateLitres, 400)
    // this.handleChange = debounce(this.handleChange, 400)
  }

  componentWillMount = () => {
    this.timer = null
  }

  componentDidUpdate = prevProps => {

    let tanks = {}
    const { data } = this.props
    if (data !== prevProps.data && data.loading === false && data.stationTanks) {
      const dips = data.dips
      fuelTypeList.forEach(ft => {

        data.stationTanks.forEach(t => {
          if (t.fuelType === ft) {
            const tmp = {
              level: '',
              litres: null,
              delivery: '',
              tank: t,
            }
            if (dips) {
              tmp.dips = R.find(R.propEq('stationTankID', t.id))(data.dips)
              tmp.level = tmp.dips.level
              tmp.litres = tmp.dips.litres
              if (tmp.dips.fuelDelivery) {
                tmp.delivery = tmp.dips.fuelDelivery.litres
              }
            }
            tanks[t.id] = tmp
          }
        })
      })

      this.setState(() => ({tanks}))
    }
  }

  // see: https://stackoverflow.com/questions/23123138/perform-debounce-in-react-js for explanation on debounce
  handleChange = (e) => {
    e.persist()
    const inputId = e.target.id
    const val     = e.target.value
    const pcs     = inputId.split('_')
    const id      = pcs[0]
    const field   = pcs[1]

    // Using a tmp tank holder appeared to be the only way to set the state properly here
    let tanks = this.state.tanks
    const value = parseInt(val, 10) || ''
    tanks[id][field] = value
    this.setState(() => ({tanks}))

    if (field === 'level') {
      this.handleCalculateLitres(id, val)
    }
  }

  // fixme: bounce is not working on this
  handleCalculateLitres = (tankID, val) => {

    // If user deletes value entirely, we need to zero out entry
    let tanks = this.state.tanks

    if (!val) {
      tanks[tankID].level = 0
      tanks[tankID].litres = 0
      this.setState(() => ({tanks}))
      return
    }

    let inputLevel = parseInt(val, 10)
    if (inputLevel <= 1) return
    const levels = tanks[tankID].tank.tank.levels
    if (!levels[inputLevel]) {
      inputLevel++
    }

    tanks[tankID].level = inputLevel
    if (levels[inputLevel]) {
      tanks[tankID].litres = levels[inputLevel].litres
      this.setState(() => ({tanks}))
    }
  }

  handleDipComplete = () => {
    this.setState({toasterMsg: 'Dip entry successfully completed'})
  }

  renderTanks = (data, classes) => {

    const { tanks } = this.state
    if (!tanks) return null

    if (data.loading) return <p>Loading...</p>

    let rows = []
    for (const t in tanks) {
      rows.push({
        id:       t,
        fuelType: tanks[t].tank.fuelType,
        tankID:   tanks[t].tank.tankID,
        size:     tanks[t].tank.tank.size,
      })
    }

    return rows.map((t, i) => (
      <div
          className={classes.dataRow}
          key={t.id}
      >
        <div className={classes.dataCell}>
          {t.size} ({t.tankID})
          <span style={{display: 'inline', marginLeft: 15}}>{t.fuelType}</span>
        </div>
        <div className={classes.dataCell}>
          <Input
              autoFocus={i === 0}
              className={classes.input}
              id={`${t.id}_level`}
              name="level"
              onChange={this.handleChange}
              type="number"
              value={tanks[t.id].level}
          />
        </div>
        <div className={classNames([classes.dataCell], [classes.dataAlignRight])}>{fmtNumber(tanks[t.id].litres, 0, true)}</div>
        <div  className={classes.dataCell}>
          <Input
              className={classes.input}
              id={`${t.id}_delivery`}
              onChange={this.handleChange}
              type="number"
              value={tanks[t.id].delivery}
          />
          </div>
      </div>
    ))
  }

  createDipsParams = () => {
    const tanks = this.state.tanks
    let error = false
    for (const t in tanks) {
      if (tanks[t].level === null) {
        error = true
        break
      }
    }
    // this.props.actions.errorSend({message: 'Generic error message here...\nfooo \n bar', type: 'danger'})

    // Extract station and date from the graphql data
    const { variables } = this.props.data
    let dips = []
    for (const stID in tanks) {
      dips.push({
        date:       variables.date,
        delivery:       tanks[stID].delivery || null,
        fuelType:       tanks[stID].tank.fuelType,
        level:          tanks[stID].level,
        litres:         tanks[stID].litres,
        stationID:  variables.stationID,
        stationTankID:  stID,
      })
    }

    if (error) {
      console.error('invalid form tanks: ', tanks) // eslint-disable-line
    }

    return { fields: dips }
  }

  fetchDipsParams = () => {
    const { variables } = this.props.data
    return {
      date:       variables.date,
      dateFrom:   variables.dateFrom,
      dateTo:     variables.dateTo,
      stationID:  variables.stationID,
    }
  }

  render() {

    const { classes, data, editMode, havePrevDayDips } = this.props
    let loading = data && data.loading ? true : false
    let submitLabel = 'Save Dips'
    if (data && data.loading === false && editMode === true) {
      submitLabel = 'Edit Dips'
    }

    return (
      <div className={classes.container}>
        {/*<Errors />*/}
        <Typography
            gutterBottom
            variant="title"
        >Dips</Typography>
        <div className={classes.headerRow}>
          <div className={classes.headerCell}>Tank</div>
          <div className={classNames([classes.headerCell], [classes.alignCenter])}>Level</div>
          <div className={classNames([classes.headerCell], [classes.alignCenter])}>Litres</div>
          <div className={classNames([classes.headerCell], [classes.alignCenter])}>Delivery</div>
          <div className={classNames([classes.headerCell], [classes.narrowCell])} />
        </div>

        {data && this.renderTanks(data, classes)}
        {data && !loading && <div>
        <Mutation
            awaitRefetchQueries
            mutation={CREATE_DIPS}
            onCompleted={this.handleDipComplete}
            refetchQueries={[{query: DIP_QUERY, variables: this.fetchDipsParams()}]}
            variables={this.createDipsParams()}
        >
          {(createDips, { loading, error }) => (
            <div className={classes.buttonContainer}>
              <Button
                  className={classes.submitButton}
                  color="primary"
                  disabled={!havePrevDayDips}
                  onClick={createDips}
                  variant="raised"
              >
                <Save className={classNames(classes.leftIcon, classes.iconSmall)} />
                {submitLabel}
              </Button>
              <div className={classes.submitMsg}>
              {loading && <div>Processing...</div>}
              {error && <div>Error :( Please try again</div>}
              </div>
            </div>
          )}
        </Mutation>
        {data && !havePrevDayDips &&
          <div>
          <Alert type="info">Previous day dips missing. Ensure dips are entered consecutively.</Alert>
          </div>
        }
        </div>
        }
        <Toaster message={this.state.toasterMsg} />
      </div>
    )
  }
}

DipForm.propTypes = {
  actions:          PropTypes.object.isRequired,
  classes:          PropTypes.object.isRequired,
  data:             PropTypes.object,
  editMode:         PropTypes.bool.isRequired,
  havePrevDayDips:  PropTypes.bool.isRequired,
}

const styles =  theme => ({
  alignRight: {
    textAlign: 'right',
  },
  alignCenter: {
    textAlign: 'center',
  },
  buttonContainer: {
    display:        'flex',
    flexDirection:  'column',
  },
  container: {
    display:        'flex',
    flexDirection:  'column',
    fontFamily:     theme.typography.fontFamily,
  },
  dataAlignRight: {
    textAlign:    'right',
    marginRight:  theme.spacing.unit * 7,
    marginLeft:   -(theme.spacing.unit * 7),
  },
  dataCell: {
    alignSelf:      'flex-end',
    flex:           1,
    padding:        theme.spacing.unit,
    paddingBottom:  theme.spacing.unit * 2,
  },
  dataRow: {
    borderBottomColor:  '#efefef',
    borderBottomStyle:  'solid',
    borderBottomWidth:  1,
    display:            'inline-flex',
    flexDirection:      'row',
    '&:hover': {
      backgroundColor: theme.palette.grey['50'],
    },
  },
  delButton: {
    minWidth: 0,
  },
  input: {
    paddingLeft:  10,
    width:        theme.spacing.unit * 15,
  },
  headerRow: {
    borderBottomColor:  '#efefef',
    borderBottomStyle:  'solid',
    borderBottomWidth:  1,
    display:            'inline-flex',
    flexDirection:      'row',
    paddingBottom:      4,
  },
  headerCell: {
    color:      theme.palette.secondary.main,
    flex:       1,
    fontWeight: '500',
    padding:    theme.spacing.unit,
  },
  leftIcon: {
    marginRight: theme.spacing.unit,
  },
  iconSmall: {
    fontSize: 20,
  },
  narrowCell: {
    flex:           .5,
    padding:        0,
    paddingBottom:  theme.spacing.unit,
  },
  submitButton: {
    margin: theme.spacing.unit,
    marginTop: theme.spacing.unit * 3,
  },
  submitMsg: {
    margin: 'auto',
  },
})

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...errorActions,
    }, dispatch),
  }
}

const StyledForm = withStyles(styles)(DipForm)

export default connect(
  null,
  mapDispatchToProps
)(StyledForm)
