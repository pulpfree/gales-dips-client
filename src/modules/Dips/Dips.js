import React, { Component } from 'react'
import PropTypes from 'prop-types'

import moment from 'moment'
import { Link } from 'react-router-dom'

import Button from '@material-ui/core/Button'
import Divider from '@material-ui/core/Divider'
import Paper from '@material-ui/core/Paper'
import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'

import DipForm from './DipForm.cntr'
import DipOverShort from './DipOverShort'
import DipSelectors from './DipSelectors.cntr'
import Header from '../Header/Header'
import Loader from '../Common/Loader'
import { datePrevDay, dateToInt } from '../../utils/date'
import { FUEL_TYPE_LIST as fuelTypeList } from '../../config/constants'

const R = require('ramda')

const populateTanks = (dips, tanks) => {
  const { curDips, prevDips } = dips
  const tanksObj = {}

  fuelTypeList.forEach((ft) => {
    tanks.forEach((t) => {
      if (t.fuelType === ft) {
        const tmp = {
          delivery: '',
          level: '',
          litres: null,
          prevLevel: '',
          tank: t,
        }
        if (curDips) {
          tmp.dips = R.find(R.propEq('stationTankID', t.id))(curDips)
          if (tmp.dips) {
            tmp.level = tmp.dips.level
            tmp.litres = tmp.dips.litres
            if (tmp.dips.fuelDelivery) {
              tmp.delivery = tmp.dips.fuelDelivery.litres
            }
          }
        }
        if (prevDips) {
          tmp.prevDips = R.find(R.propEq('stationTankID', t.id))(prevDips)
          tmp.prevLevel = tmp.prevDips && tmp.prevDips.level
        }
        tanksObj[t.id] = tmp
      }
    })
  })

  return tanksObj
}

class Dips extends Component {
  constructor(props) {
    super(props)

    this.state = {
      submitting: false,
    }
    this.handleSubmitNotify = this.handleSubmitNotify.bind(this)
  }

  componentDidUpdate = (prevProps) => {
    if (!prevProps.dips) return
    if (!prevProps.dips.error && this.props.dips.error) {
      const errMsg = this.props.dips.error.message
      const ts = moment.utc().format()
      const msg = `${errMsg} -- ${ts}`
      this.props.actions.errorSend({ message: msg, type: 'danger' })
    }
  }

  handleSubmitNotify = (isSubmit) => {
    this.setState(() => ({ submitting: isSubmit }))
  }

  render() {
    const {
      classes,
      dips,
      fuelSaleLatest,
      history,
      match: { params },
      tanks,
    } = this.props

    const requestDate = params.date
    const dateObj = moment(requestDate)
    let havePrevDayDips = false
    let editMode = false
    let haveFSImport = true
    let loading = true

    let fsDate = ''
    if (fuelSaleLatest && fuelSaleLatest.fuelSaleLatest) {
      fsDate = moment(fuelSaleLatest.fuelSaleLatest.date.toString())
    }

    // Compare last imported with current date
    if (dateObj.isAfter(fsDate)) {
      haveFSImport = false
    }

    let waitComponent
    if (!params.stationID) {
      waitComponent = <div>Select a Station</div>
    } else {
      waitComponent = <Loader />
    }

    if (dips && dips.networkStatus === 7
      && tanks && tanks.networkStatus === 7) {
      loading = false
    }

    if (dips && dips.dipOverShortRange) {
      const curDay = dateToInt(requestDate)
      const prevDay = datePrevDay(requestDate)
      if (dips.dipOverShortRange[0].date === prevDay) {
        havePrevDayDips = true
      }
      if (dips.dipOverShortRange[1] && dips.dipOverShortRange[1].date === curDay) {
        editMode = true
      }
    }

    return (
      <div>
        <Header history={history} />
        <Paper className={classes.container}>
          <Typography
            gutterBottom
            variant="h5"
          >Dip Entries
          </Typography>
          <Divider /><br />
          <div className={classes.secondaryContainer}>
            <DipSelectors />

            {haveFSImport && loading && waitComponent}

            {!haveFSImport &&
              <div className={classes.lastFS}>
                <Button
                  className={classes.lastFSButton}
                  color="secondary"
                  component={Link}
                  size="small"
                  to="/import-data"
                  variant="outlined"
                >
                  Import Data
                </Button>
                <br />Date of last Fuel Sale import: {fsDate.format('MMM DD, YYYY')}
              </div>
            }

            {haveFSImport && !loading &&
              <div style={{ display: 'flex', flexDirection: 'row', marginTop: 20 }}>
                <div style={{ flex: 1 }}>
                  <DipForm
                    editMode={editMode}
                    havePrevDayDips={havePrevDayDips}
                    isSubmit={this.state.submitting}
                    tankDips={populateTanks(dips, tanks.stationTanks)}
                    submitFunc={this.handleSubmitNotify}
                  />
                </div>
                <div style={{ flex: 0.7 }}>
                  <DipOverShort
                    dateObj={dateObj}
                    dips={dips}
                    isSubmit={this.state.submitting}
                  />
                </div>
              </div>
            }

          </div>
        </Paper>
      </div>
    )
  }
}
Dips.propTypes = {
  actions: PropTypes.instanceOf(Object).isRequired,
  classes: PropTypes.instanceOf(Object).isRequired,
  dips: PropTypes.instanceOf(Object),
  fuelSaleLatest: PropTypes.instanceOf(Object),
  history: PropTypes.instanceOf(Object).isRequired,
  match: PropTypes.instanceOf(Object).isRequired,
  tanks: PropTypes.instanceOf(Object),
}
Dips.defaultProps = {
  dips: null,
  fuelSaleLatest: null,
  tanks: null,
}

const styles = theme => ({
  container: {
    fontFamily: theme.typography.fontFamily,
    minHeight: 600,
    margin: theme.spacing.unit * 2,
    padding: theme.spacing.unit * 2,
  },
  lastFS: {
    color: theme.palette.secondary.main,
    marginTop: theme.spacing.unit,
    fontSize: 14,
  },
  lastFSButton: {
    marginBottom: theme.spacing.unit,
  },
  secondaryContainer: {
    display: 'flex',
    flex: 1,
    flexDirection: 'column',
    margin: 'auto',
    width: 1200,
  },
})

export default withStyles(styles)(Dips)
