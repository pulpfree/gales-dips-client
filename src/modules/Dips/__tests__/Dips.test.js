/* eslint no-undef: "off" */

import React from 'react'
import { shallow } from 'enzyme'

import Dips from '../Dips.cntr'

describe('Dips', () => {

  let mountedC
  let props
  const reportC = () => {
    if (!mountedC) {
      mountedC = shallow(
        <Dips {...props} />
      )
    }
    return mountedC
  }

  beforeEach(() => {
    props = {
      history: {},
      location: {pathname: '/'},
      match: {
        params: {
          date: '',
        },
      },
    }
    mountedC = undefined
  })

  it('matches snapshot', () => {
    expect(reportC()).toMatchSnapshot()
  })

})