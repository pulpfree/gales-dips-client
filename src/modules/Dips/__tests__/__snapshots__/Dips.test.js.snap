// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dips matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MemoryRouter>
    <Apollo(Connect(WithStyles(Dips)))
        history={Object {}}
        location={
            Object {
                "pathname": "/",
              }
        }
        match={Object {}}
    />
</MemoryRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Apollo(Connect(WithStyles(Dips)))
        history={Object {}}
        location={
                Object {
                        "pathname": "/",
                      }
        }
        match={Object {}}
/>,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "lj9v53",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "lj9v53",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "history": Object {},
        "location": Object {
          "pathname": "/",
        },
        "match": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Apollo(Connect(WithStyles(Dips)))
          history={Object {}}
          location={
                    Object {
                              "pathname": "/",
                            }
          }
          match={Object {}}
/>,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "lj9v53",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "lj9v53",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "history": Object {},
          "location": Object {
            "pathname": "/",
          },
          "match": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
